java/org/apache/catalina/realm/DataSourceRealm.java

```diff
@@ -510,21 +582,22 @@ protected Principal getPrincipal(String username) {
             return null;
         }
 
-        ArrayList<String> list = null;
+        // Using a Set removes duplicate roles
+        Set<String> roles = null;
 
         try (PreparedStatement stmt = dbConnection.prepareStatement(preparedRoles)) {
             stmt.setString(1, username);
 
             try (ResultSet rs = stmt.executeQuery()) {
-                list = new ArrayList<>();
+                roles = new HashSet<>();
```

Comment:
Don't change this here. It has nothing to do with the new feature.

Reply:
OK